### PR TITLE
Add node browser

### PR DESCRIPTION
**Pull request by fakob**
* Added a node browser which
  * shows all nodes in a graph
    * nodes with errors and macros are on top
  * hovering over a node in the list shows a focus border around the node in the graph
  * clicking on a node in the list centers the node
  * double clicking on a node in the list centers and zooms in onto the node
  * clicking the select node button on a node in the list, selects the node
  * using the arrow in the top of the drawer, let's you go back to the node browser
  * the INFO tab shows
    * a description we can add in the future
    * the JSON of the graph which can be edited and with the replace button be loaded again
* Changed the default name of a node on creation to NOT be the node type, but the given name (getName())
* Updated the "missing nodes test" example

https://github.com/fakob/plug-and-play/assets/4619772/c8687cd4-7bcc-421a-a86f-94ac6824ce41

### CODE REVIEW

File: src/NodeArrayContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+      item.name.toLowerCase().includes(filter) ||
+      item.type.toLowerCase().includes(filter) ||
+      item.id.toLowerCase().includes(filter)
+    );
+  };
+
+  // Custom sort function to sort items alphabetically
+  const customSort = (a, b) => {
+    // order nodes with errors on top
+    if (!a.successfullyExecuted && b.successfullyExecuted) return -1;
```
Comment (magnificus) [inline src/NodeArrayContainer.tsx:453]: This sorting can be described more elegantly for example via something like

Return 100 * (a.successexec - bsuccessexec) + 10 * (a.type == macro ...) ...

File: src/NodeArrayContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+    PPGraph.currentGraph?.nodes !== undefined &&
+      Object.keys(PPGraph.currentGraph?.nodes).length,
+  ]);
+
+  useEffect(() => {
+    setConfigData(getConfigData(PPGraph.currentGraph));
+  }, [PPGraph.currentGraph?.id]);
+
+  useEffect(() => {
+    const newSelectedNode =
```
Comment (magnificus) [inline src/NodeArrayContainer.tsx:500]: These long and anxious statements are a bit confusing and ugly, code should know what is coming in here

File: src/NodeArrayContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+      />
+      <Stack
+        spacing={1}
+        sx={{
+          mt: 1,
+          overflow: 'auto',
+          height: 'calc(100vh - 120px)',
+        }}
+      >
+        {(props.filter === 'nodes' || props.filter == null) && (
```
Comment (magnificus) [inline src/NodeArrayContainer.tsx:525]: Same with all these null checks

**Review by magnificus — CHANGES_REQUESTED** [review]
It looks very promising and looking forward to trying it, but some mess in the code we should do a second sweep on

File: src/GraphInspectorContainer.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  filter: string;
+  setFilter: React.Dispatch<React.SetStateAction<string>>;
+  filterText: string;
+  setFilterText: React.Dispatch<React.SetStateAction<string>>;
+};
+
+const GraphInspectorContainer: React.FunctionComponent<
+  GraphInspectorContainerProps
+> = (props) => {
+  const [graphName, setGraphName] = React.useState('');
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/GraphInspectorContainer.tsx:53]: If I should pass it as a prop here I would need to get it somewhere up the chain and pass it down. But then I need the state somewhere up there. I don't think I can move the function to get the name into a prop, as it returns a promise.

File: src/InspectorContainer.tsx [hunk 5]
```diff
@@ -64,7 +83,25 @@ function InspectorHeader(props) {
...
+            color: `${
+              Color(props.randomMainColor).isDark()
+                ? COLOR_WHITE_TEXT
+                : COLOR_DARK
+            }`,
+          }}
+          onClick={() => {
+            PPGraph.currentGraph.selection.deselectAllNodes();
+          }}
+        >
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/InspectorContainer.tsx:100]: You are right, I was lazy when I had done that. Will fix

File: src/InspectorContainer.tsx [hunk 6]
```diff
@@ -144,9 +188,7 @@ type InspectorContainerProps = {
 const InspectorContainer: React.FunctionComponent<InspectorContainerProps> = (
   props,
 ) => {
-  const [nodeName, setNodeName] = React.useState(
-    props.selectedNodes?.[0]?.name,
-  );
+  const [nodeName, setNodeName] = React.useState('');
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/InspectorContainer.tsx:192]: I need a state when editing so the updating works, but I can move it into the header component now

File: src/InspectorContainer.tsx [hunk 7]
```diff
@@ -117,7 +161,7 @@ function InspectorHeader(props) {
           <EditIcon fontSize="inherit" />
         </IconButton>
       </Box>
-      {props.selectedNodes[0].type !== props.nodeName && (
+      {props.selectedNodes[0].name !== props.selectedNodes[0].getName() && (
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/InspectorContainer.tsx:164]: will do

File: src/NodeArrayContainer.tsx [hunk 8]
```diff
@@ -0,0 +1,593 @@
...
+      }}
+      onClick={(event: React.MouseEvent<HTMLLIElement>) => {
+        event.stopPropagation();
+        const nodeToJumpTo = PPGraph.currentGraph.nodes[props.property.id];
+        if (nodeToJumpTo) {
+          ensureVisible([nodeToJumpTo]);
+          setTimeout(() => {
+            PPGraph.currentGraph.selection.drawSingleFocus(nodeToJumpTo);
+          }, 800);
+          if (event.detail === 2) {
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/NodeArrayContainer.tsx:137]: Good point. It refers to double and tripple click on an onclick event. Will do

File: src/NodeArrayContainer.tsx [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+  setFilter: React.Dispatch<React.SetStateAction<string>>;
+  filterText: string;
+  setFilterText: React.Dispatch<React.SetStateAction<string>>;
+};
+
+export const NodeArrayContainer: React.FunctionComponent<
+  NodeArrayContainerProps
+> = (props) => {
+  const singleNode = props.selectedNodes.length ? props.selectedNodes[0] : null;
+  const [selectedNode, setSelectedNode] = useState(singleNode);
```
[the remark this answers is not in the document]
Comment (fakob) [reply src/NodeArrayContainer.tsx:405]: Actually, good catch. The selectedNode is not needed here at all. Was a leftover from before. Will remove.

File: src/NodeArrayContainer.tsx (hunk 1 above)
Comment (fakob) [reply src/NodeArrayContainer.tsx:453]: Ah, good idea.

File: src/NodeArrayContainer.tsx (hunk 2 above)
Comment (fakob) [reply src/NodeArrayContainer.tsx:500]: removed it, as mentioned above

File: src/NodeArrayContainer.tsx (hunk 3 above)
Comment (fakob) [reply src/NodeArrayContainer.tsx:525]: Sounds good

File: src/GraphInspectorContainer.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (magnificus) [reply src/GraphInspectorContainer.tsx:53]: but we are using graph name in other components as well right? Could they refer to the same place?

**Review by magnificus — APPROVED** [review]

**Pull request merged**